packages/sdk/identity/src/odis/query.ts

```diff
@@ -129,44 +120,55 @@ export function signWithDEK(message: string, signer: EncryptionKeySigner) {
  */
 export async function queryOdis<ResponseType>(
   signer: AuthSigner,
-  body: PhoneNumberPrivacyRequest,
+  body: PhoneNumberPrivacyRequest | DomainRequest<KnownDomain>,
   context: ServiceContext,
   endpoint: string
-) {
+): Promise<ResponseType> {
   debug(`Posting to ${endpoint}`)
 
-  // Sign payload using account privkey
   const bodyString = JSON.stringify(body)
 
-  let authHeader = ''
+  // Sign payload using provided account and authentication method.
+  // NOTE: Signing and verifying signatures over JSON encoded blobs relies on both the serializer
```

Comment:
Great point. Would a better approach be to parse out the fields and sign them according to a function exposed in phone-number-privacy-common?

Reply:
So the main result of this is that anyone who wants to verify the signature over this data must not decode and re-encode it (i.e. they need to save and pass along the original encoded string). This is a situation that seems to be pretty common, including in standards like JWTs, and is not deadly but is unfortunate. I get the sense from what I've read on this topic that its not uncommon for production systems to simply accept the requirement of passing around the originally encoded string. (I was actually talking to my friend this morning about a [AWS Sigv4](https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html) which has this issue to an extent)

One solution that gets deployed is deterministic encoding rules which are restricted versions of the original encoding rules. Examples include X.509 DER which is a subset of X.509 BER and is used in certificates, [OJSON](https://github.com/potatosalad/erlang-ojson), and [CBOR core deterministic encoding requirements](https://www.rfc-editor.org/rfc/rfc8949.html#name-core-deterministic-encoding).

In CIP-40, the solution I chose is EIP-712 which provides deterministic "encoding" specifically for the purposes of hashing and signing. With that, the idea is that you can encode, decode and pass around the data to your hearts content and ensure that when its reassembled you will get the same hash. (It also ensures domain seperation to prevent the tail-risk of some colliding message format)

I wouldn't rush to change anything, although if we face any problem with this we may consider using EIP-712 in the future.

Reply:
Awesome! Thanks for the detailed explanation. Using eip712 makes a lot of sense